File: docs/lib/collection_catalogue.js
import { CatalogueConnector } from "./abstractions/catalogue_connector.js";
import { CatalogueFilters } from "./catalogue_filters.js";
import { Page } from "./page.js";
import * as CatalogueStatics from "./catalogue_statics.js";
import { Result } from "./result.js";
import * as PageUtils from "./utils/page_utils.js";

/**
 * @template TItem 
 * @template TFilters
 * @implements {CatalogueConnector}
 */
class CollectionCatalogue {

    /** @type {TItem[]} */
    collection;

    /** @type {string[]} */
    errors;

    /**
     * @param {TItem[]} collection 
     */
    constructor(collection) {
        this.collection = collection;
        this.errors = [];
    }

    /** @returns {boolean} */
    get available() { return true; }

    /**
     * @param {TItem[]} items 
     * @param {TFilters} filters 
     * @returns {TItem[]}
     */
    queryItems(items, filters) { return items; }

    /**
     * @param {TItem} item 
     * @returns {boolean}
     */
    validateInsert(item) { return true; }

    /**
     * @param {TItem} item 
     * @returns {boolean}
     */
    validateUpdate(item) { return true; }

    /**
     * @param {TItem} item 
     * @returns {boolean}
     */
    validateDelete(item) { return true; }

    /**
     * @param {TFilters} filters 
     * @returns {Promise<Result<Page<TItem>>>}
     */
    async pageAsync(filters = null) {
        this.errors = [];
        let items = this.collection;
        if (filters?.ids?.length) items = items.filter(x => filters.ids.includes(x.id));
        if (filters != null) items = this.queryItems(items, filters);
        let page = PageUtils.getPage(items, filters?.pageIndex ?? 0, filters?.pageSize ?? CatalogueStatics.DefaultChunkSize);
        return new Result({
            data: new Page({
                items: page.items,
                pageIndex: page.pageIndex,
                pageCount: page.pageCount,
                changesAt: new Date()
            })
        });
    }

    /**
     * @param {string} id 
     * @returns {Promise<Result<TItem>>}
     */
    async getAsync(id) {
        this.errors = [];
        let item = this.collection.find(x => x.id == id);
        return new Result({
            data: item
        });
    }

    /**
     * @param {TItem} item 
     * @returns {Promise<Result<string>>}
     */
    async insertAsync(item) {
        this.errors = [];
        if (item.validate && !item.validate(this.errors)) return new Result({ errors: this.errors });
        if (this.collection.some(x => x.id == item.id)) {
            this.errors.push(CatalogueStatics.RepeatedItem);
            return new Result({ errors: this.errors });
        }
        if (!this.validateInsert(item)) return new Result({ errors: this.errors });
        this.collection.push(item);
        return new Result({ data: item.id });
    }

    /**
     * @param {string} id 
     * @param {TItem} item 
     * @returns {Promise<Result<boolean>>}
     */
    async updateAsync(id, item) {
        this.errors = [];
        if (item.validate && !item.validate(this.errors)) return new Result({ errors: this.errors });
        let original = this.collection.find(x => x.id == item.id);
        if (original == null) {
            this.errors.push(CatalogueStatics.NoItem);
            return new Result({ errors: this.errors });
        }
        if (item.UpdatedAt < original.UpdatedAt) return new Result({ data: false });
        if (!this.validateUpdate(item)) return new Result({ errors: this.errors });
        original.update(item);
        return new Result({ data: true });
    }

    /**
     * @param {string} id 
     * @returns {Promise<Result<boolean>>}
     */
    async deleteAsync(id) {
        this.errors = [];
        let item = this.collection.find(x => x.id == item.id);
        if (item == null) {
            this.errors.push(CatalogueStatics.NoItem);
            return new Result({ errors: this.errors });
        }
        if (!this.validateDelete(item)) return new Result({ errors: this.errors });
        this.collection.splice(this.collection.indexOf(item), 1);
        return new Result({ data: true });
    }

}

export {
    CollectionCatalogue
}

File: Core/CollectionCatalogue.cs
using KolibSoft.Catalogue.Core.Abstractions;
using KolibSoft.Catalogue.Core.Utils;

namespace KolibSoft.Catalogue.Core;

public class CollectionCatalogue<TItem, TFilters> : ICatalogueConnector<TItem, TFilters>
    where TItem : IItem, IValidatable, IUpdatable<TItem>
    where TFilters : IPageFilters, IItemFilters
{

    public ICollection<TItem> Collection { get; }
    public ICollection<string> Errors { get; } = new List<string>();
    public virtual bool Available => true;

    protected virtual IQueryable<TItem> QueryItems(IQueryable<TItem> items, TFilters? filters = default) => items;
    protected virtual bool ValidateInsert(TItem item) => true;
    protected virtual bool ValidateUpdate(TItem item) => true;
    protected virtual bool ValidateDelete(TItem item) => true;

    public virtual Task<Result<Page<TItem>?>> PageAsync(TFilters? filters = default) => Task.Run<Result<Page<TItem>?>>(() =>
    {
        Errors.Clear();
        var items = Collection.AsQueryable();
        if (filters?.UpdatesAt != null) items = items.Where(x => x.UpdatedAt >= filters.UpdatesAt);
        if (filters?.Ids?.Any() == true) items = items.Where(x => filters.Ids.Contains(x.Id));
        items = QueryItems(items, filters);
        var page = items.GetPage(filters?.PageIndex ?? 0, filters?.PageSize ?? CatalogueStatics.DefaultChunkSize);
        return page;
    });

    public virtual Task<Result<TItem?>> GetAsync(Guid id) => Task.Run<Result<TItem?>>(() =>
    {
        Errors.Clear();
        var item = Collection.FirstOrDefault(x => x.Id == id);
        if (item == null)
        {
            Errors.Add(CatalogueStatics.NoItem);
            return Errors.ToArray();
        }
        return item;
    });

    public virtual Task<Result<Guid?>> InsertAsync(TItem item) => Task.Run<Result<Guid?>>(() =>
    {
        Errors.Clear();
        if (!item.Validate(Errors)) return Errors.ToArray();
        if (Collection.Any(x => x.Id == item.Id))
        {
            Errors.Add(CatalogueStatics.RepeatedItem);
            return Errors.ToArray();
        }
        if (!ValidateInsert(item)) return Errors.ToArray();
        Collection.Add(item);
        return item.Id;
    });

    public virtual Task<Result<bool?>> UpdateAsync(Guid id, TItem item) => Task.Run<Result<bool?>>(() =>
    {
        Errors.Clear();
        if (!item.Validate(Errors)) return Errors.ToArray();
        var original = Collection.FirstOrDefault(x => x.Id == id);
        if (original == null)
        {
            Errors.Add(CatalogueStatics.NoItem);
            return Errors.ToArray();
        }
        if (item.UpdatedAt < original.UpdatedAt) return false;
        if (!ValidateUpdate(item)) return Errors.ToArray();
        original.Update(item);
        return true;
    });

    public virtual Task<Result<bool?>> DeleteAsync(Guid id) => Task.Run<Result<bool?>>(() =>
    {
        Errors.Clear();
        var item = Collection.FirstOrDefault(x => x.Id == id);
        if (item == null)
        {
            Errors.Add(CatalogueStatics.NoItem);
            return Errors.ToArray();
        }
        if (!ValidateDelete(item)) return Errors.ToArray();
        Collection.Remove(item);
        return true;
    });

    public CollectionCatalogue(ICollection<TItem> collection)
    {
        Collection = collection;
    }

}